File: sp-2 security/SCM/src/main/java/com/example/SCM/entity/QCInspection.java
package com.example.SCM.entity;

import com.example.SCM.enumClass.ResultStatus;
import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "qc_inspections")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class QCInspection {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @JsonIgnore
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "grn_id", nullable = false)
    private GoodsReceivedNote goodsReceivedNote;

    @JsonIgnore
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "product_id", nullable = false)
    private Product product;

    @JsonIgnore
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "inspected_by", nullable = false)
    private User inspectedBy;

    @Column(nullable = false)
    private String inspectionType;

    private int sampleSize;

    private int defectsFound;

    @Column(columnDefinition = "TEXT")
    private String defectDescription;

    @Enumerated(EnumType.STRING)
    private ResultStatus result;


    private String certificateRef;


    private String labTestReport;

    @Column(nullable = false)
    private LocalDate inspectedAt;

    @JsonIgnore
    @OneToMany(mappedBy = "qcInspection", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    private List<QCChecklist> checklists = new ArrayList<>();

    @Column(updatable = false, nullable = false)
    private LocalDateTime createdAt;

    @Column(nullable = false)
    private LocalDateTime updatedAt;

    @PrePersist
    protected void onCreate() {
        this.createdAt = LocalDateTime.now();
        this.updatedAt = LocalDateTime.now();
        if (this.inspectedAt == null) {
            this.inspectedAt = LocalDate.now();
        }
    }

    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = LocalDateTime.now();
    }
}

File: Anguler/SCM_Anguler/src/app/component/shared/model/qc-inspection.ts

export interface QCChecklistRequestModel {
  inspectionId?: number; 
  checkpointName: string;
  isPassed: boolean;
  remarks: string;
}


export interface QCChecklistResponseModel {
  id: number; 
  checkpointName: string;
  isPassed: boolean;
  remarks: string;
  createdAt: string; 
  updatedAt: string;
  inspectionId: number;
  inspectionType: string;
}


export interface QCInspectionRequestModel {
  id?: number; 
  grnId: number;
  productId: number;
  inspectionType: string; // e.g., 'VISUAL', 'LAB_TEST'
  inspectedBy: number;
  sampleSize: number;
  defectsFound: number;
  defectDescription: string;
  result: 'GOOD' | 'VERY_GOOD' | 'AVERAGE' | 'BAD' | string;
  certificateRef: string;
  labTestReport: string;
  inspectedAt: string; 
  checklists: QCChecklistRequestModel[]; 
}


export interface QCInspectionResponseModel {
  id: number;
  inspectionType: string;
  sampleSize: number;
  defectsFound: number;
  defectDescription: string;
  result: 'GOOD' | 'VERY_GOOD' | 'AVERAGE' | 'BAD' | string;
  certificateRef: string;
  labTestReport: string;
  inspectedAt: string; // YYYY-MM-DD
  createdAt: string;
  updatedAt: string;
  
  
  grnId: number;
  grnNumber: string;
  productId: number;
  productName: string;
  inspectedBy: number;
  inspectedByName: string;
  
  checklists: QCChecklistResponseModel[]; 
}
